From 07f06cad8b7389952ce885fe37cb26f1b9f32483 Mon Sep 17 00:00:00 2001
From: Vinayakumar B <vinayakumarb@apache.org>
Date: Wed, 17 Feb 2016 10:13:41 +0530
Subject: [PATCH 1251/2848] HADOOP-12810. FileSystem#listLocatedStatus causes
 unnecessary RPC calls (Contributed by Ryan Blue)

(cherry picked from commit 96ea3094315bb1e1a5e268e3817c7fdedc3e9462)
(cherry picked from commit 2bf5a005fab287689a148c1928f3b04c0a8dcedf)
(cherry picked from commit a9c1bb42418dbde8958dbbc115f6045e696ea2ea)

Change-Id: I7b7c17608e43020ca013da025a68cb61d631bef3
---
 .../main/java/org/apache/hadoop/fs/FileSystem.java |    4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/FileSystem.java b/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/FileSystem.java
index 3d4221b..70a29af 100644
--- a/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/FileSystem.java
+++ b/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/FileSystem.java
@@ -1705,8 +1705,10 @@ public LocatedFileStatus next() throws IOException {
           throw new NoSuchElementException("No more entry in " + f);
         }
         FileStatus result = stats[i++];
+        // for files, use getBlockLocations(FileStatus, int, int) to avoid
+        // calling getFileStatus(Path) to load the FileStatus again
         BlockLocation[] locs = result.isFile() ?
-            getFileBlockLocations(result.getPath(), 0, result.getLen()) :
+            getFileBlockLocations(result, 0, result.getLen()) :
             null;
         return new LocatedFileStatus(result, locs);
       }
-- 
1.7.9.5

